samza-core/src/main/java/org/apache/samza/container/grouper/stream/AbstractSystemStreamPartitionGrouper.java

```diff
@@ -0,0 +1,186 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.samza.container.grouper.stream;
+
+import com.google.common.base.Preconditions;
+import org.apache.samza.annotation.InterfaceStability;
+import org.apache.samza.config.Config;
+import org.apache.samza.config.JobConfig;
+import org.apache.samza.config.TaskConfigJava;
+import org.apache.samza.container.TaskName;
+import org.apache.samza.metrics.MetricsRegistryMap;
+import org.apache.samza.system.SystemStream;
+import org.apache.samza.system.SystemStreamPartition;
+import org.apache.samza.util.Util;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import java.util.*;
+
+/**
+ * An abstract implementation of {@link SystemStreamPartitionGrouper} that provides a stream expansion
+ * aware task to partition assignments.
+ */
+@InterfaceStability.Evolving
+public abstract class AbstractSystemStreamPartitionGrouper implements SystemStreamPartitionGrouper {
```

Comment:
Are all SSP grouper implementations expected to extend this? If so, document it. If not, why make this an abstract class instead of using the functionality here as util methods in implementations that need it?

Reply:
An easier way to implement this very same logic maybe to just enrich the existing ```GroupBySystemStreamPartition```, and have it take the GrouperContext as an optional constructor input.

Reply:
In that case s/Abstract/Base to clarify that this is the expected base class for all groupers. Also, there are no abstract methods in this class, what are the extending classes expected to implement?

Reply:
The `SystemStreamPartitionGrouper` interface doesn't mention this. There's also nothing preventing them from overriding this method. If this functionality is 
1. going to be common to and required for all grouper implementations, and
2. provided by the framework
it will be better to use composition over inheritence, and use the user provided grouper and mapper implementation within this class. 
This will also address @rmatharu's comments about adding a new method to the grouper and moving some of this logic to the default method implementation - it will be unnecessary. Does that sound reasonable?

Reply:
Done.